solr/core/src/java/org/apache/solr/response/transform/ChildDocTransformer.java

```diff
@@ -0,0 +1,263 @@
+
+  private final String name;
+  private final BitSetProducer parentsFilter;
+  private final DocSet childDocSet;
+  private final int limit;
+  private final boolean isNestedSchema;
+
+  private final SolrReturnFields childReturnFields = new SolrReturnFields();
+
+  ChildDocTransformer(String name, BitSetProducer parentsFilter,
+                      DocSet childDocSet, boolean isNestedSchema, int limit) {
+    this.name = name;
+    this.parentsFilter = parentsFilter;
+    this.childDocSet = childDocSet;
+    this.limit = limit;
+    this.isNestedSchema = isNestedSchema;
+  }
+
+  @Override
+  public String getName()  {
+    return name;
+  }
+
+  @Override
+  public void transform(SolrDocument rootDoc, int rootDocId) {
+    // note: this algorithm works if both if we have have _nest_path_  and also if we don't!
+
+    try {
+
+      // lookup what the *previous* rootDocId is, and figure which segment this is
+      final SolrIndexSearcher searcher = context.getSearcher();
+      final List<LeafReaderContext> leaves = searcher.getIndexReader().leaves();
+      final int seg = ReaderUtil.subIndex(rootDocId, leaves);
+      final LeafReaderContext leafReaderContext = leaves.get(seg);
+      final int segBaseId = leafReaderContext.docBase;
+      final int segRootId = rootDocId - segBaseId;
+      final BitSet segParentsBitSet = parentsFilter.getBitSet(leafReaderContext);
+
+      final int segPrevRootId = segRootId==0? -1: segParentsBitSet.prevSetBit(segRootId - 1); // can return -1 and that's okay
+
+      if(segPrevRootId == (segRootId - 1)) {
+        // doc has no children, return fast
+        return;
+      }
+
+      // we'll need this soon...
+      final SortedDocValues segPathDocValues = DocValues.getSorted(leafReaderContext.reader(), NEST_PATH_FIELD_NAME);
+      // passing a different SortedDocValues obj since the child documents which come after are of smaller docIDs,
+      // and the iterator can not be reversed.
+      final String transformedDocPath = getPathByDocId(segRootId, DocValues.getSorted(leafReaderContext.reader(), NEST_PATH_FIELD_NAME));
+
+      // the key in the Map is the document's ancestors key(one above the parent), while the key in the intermediate
+      // MultiMap is the direct child document's key(of the parent document)
+      Map<String, Multimap<String, SolrDocument>> pendingParentPathsToChildren = new HashMap<>();
+
+      SolrDocumentFetcher docFetcher = searcher.getDocFetcher();
+      final int lastChildId = segBaseId + segPrevRootId + 1;
+      // Loop each child ID up to the parent (exclusive).
+      for (int docId = calcDocIdToIterateFrom(lastChildId, rootDocId); docId < rootDocId; ++docId) {
+
+        // get the path.  (note will default to ANON_CHILD_KEY if schema is not nested or empty string if blank)
+        String fullDocPath = getPathByDocId(docId - segBaseId, segPathDocValues);
+
+        if(isNestedSchema && !fullDocPath.contains(transformedDocPath)) {
```

Comment:
Perhaps a better way to do this would be building a new Filter for every transformed doc e.g. `_nest_path_:transformedDocPath`?
I am not quite sure of the performance overhead such technique would impose,
WDYT?

Reply:
> Perhaps a better way to do this ...

I think it would be slow and add complexity if we did that.

string.contains(...) doesn't seem right here; shouldn't it be startsWith or endsWith?  Will probably need to special-case empty transformedDocPath (rootDocPath).

I think it's possible to not need "isNestedSchema", though it's not a big deal.